# Data Cleaning Challenge  : Cleaning the FIFA21 dataset scraped and Available on [KAGGLE](https://www.kaggle.com/datasets/yagunnersya/fifa-21-messy-raw-dataset-for-cleaning-exploring "KAGGLE FIFA21 dataset")

![FIFA2](FIFA21.png)

### Importing basic libraries

In [249]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

### Importing additif library

In [278]:
# importing the re library for handling regular expressions
import re

## load the dataset and preliminary explorations

In [279]:
# load the dataset1 (fifa21_raw_data)
df1= pd.read_csv('fifa21_raw_data.csv')
df1.shape

C:\ProgramData\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3165: DtypeWarning: Columns (76) have mixed types.Specify dtype option on import or set low_memory=False.
  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,


(18979, 77)

In [280]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18979 entries, 0 to 18978
Data columns (total 77 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   photoUrl          18979 non-null  object
 1   LongName          18979 non-null  object
 2   playerUrl         18979 non-null  object
 3   Nationality       18979 non-null  object
 4   Positions         18979 non-null  object
 5   Name              18979 non-null  object
 6   Age               18979 non-null  int64 
 7   ↓OVA              18979 non-null  int64 
 8   POT               18979 non-null  int64 
 9   Team & Contract   18979 non-null  object
 10  ID                18979 non-null  int64 
 11  Height            18979 non-null  object
 12  Weight            18979 non-null  object
 13  foot              18979 non-null  object
 14  BOV               18979 non-null  int64 
 15  BP                18979 non-null  object
 16  Growth            18979 non-null  int64 
 17  Joined      

In [282]:
# load the dataset2 (fifa21 raw data v2)
df2 = pd.read_csv('fifa21 raw data v2.csv')
df2.shape

(18979, 77)

In [254]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18979 entries, 0 to 18978
Data columns (total 77 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   ID                18979 non-null  int64 
 1   Name              18979 non-null  object
 2   LongName          18979 non-null  object
 3   photoUrl          18979 non-null  object
 4   playerUrl         18979 non-null  object
 5   Nationality       18979 non-null  object
 6   Age               18979 non-null  int64 
 7   ↓OVA              18979 non-null  int64 
 8   POT               18979 non-null  int64 
 9   Club              18979 non-null  object
 10  Contract          18979 non-null  object
 11  Positions         18979 non-null  object
 12  Height            18979 non-null  object
 13  Weight            18979 non-null  object
 14  Preferred Foot    18979 non-null  object
 15  BOV               18979 non-null  int64 
 16  Best Position     18979 non-null  object
 17  Joined      

In [284]:
# Overview of the dataset
df1.sample(5)

photoUrl  \
2555   https://cdn.sofifa.com/players/135/190/21_60.png   
7235   https://cdn.sofifa.com/players/186/595/21_60.png   
12536  https://cdn.sofifa.com/players/244/106/21_60.png   
16534  https://cdn.sofifa.com/players/238/009/21_60.png   
8603   https://cdn.sofifa.com/players/252/704/21_60.png   

                     LongName  \
2555           Michael Liendl   
7235          Elliott Bennett   
12536  José Pascual Alba Seva   
16534            Martin Frese   
8603               Marco Sala   

                                               playerUrl Nationality  \
2555   http://sofifa.com/player/135190/michael-liendl...     Austria   
7235   http://sofifa.com/player/186595/elliott-bennet...     England   
12536  http://sofifa.com/player/244106/jose-pascual-a...       Spain   
16534  http://sofifa.com/player/238009/martin-frese/2...     Denmark   
8603   http://sofifa.com/player/252704/marco-sala/210...       Italy   

      Positions        Name  Age  ↓OVA  POT  \
2555     CAM CM   M. Liendl   34    73   73   
7235      RB LB  E. Bennett   31    68   68   
12536     CM RB       Pascu   20    63   77   
16534     LB LW    M. Frese   22    58   67   
8603         LB     M. Sala   21    66   78   

                                 Team & Contract  ...     A/W     D/W   IR  \
2555     \n\n\n\nWolfsberger AC\n2018 ~ 2021\n\n  ...  Medium  Medium  1 ★   
7235   \n\n\n\nBlackburn Rovers\n2016 ~ 2021\n\n  ...    High    High  1 ★   
12536      \n\n\n\nADO Den Haag\n2020 ~ 2021\n\n  ...  Medium  Medium  1 ★   
16534   \n\n\n\nFC Nordsjælland\n2017 ~ 2022\n\n  ...    High  Medium  1 ★   
8603               \n\n\n\nSPAL\n2019 ~ 2024\n\n  ...  Medium  Medium  1 ★   

      PAC  SHO PAS  DRI DEF PHY Hits  
2555   54   72  77   72  52  66  \n3  
7235   73   61  66   67  64  66  \n9  
12536  56   56  64   63  56  44  \n4  
16534  73   50  52   62  49  58    3  
8603   72   33  52   64  63  58  \n8  

[5 rows x 77 columns]

In [285]:
df2.sample(5)

ID          Name           LongName  \
2694   206427      B. Butko       Bogdan Butko   
4026   243019     T. Băluță       Tudor Băluță   
5235   214993   P. Alvarado     Pablo Alvarado   
18689  256442     J. Matete         Jay Matete   
10221  246946  I. Coulibaly  Ismaila Coulibaly   

                                               photoUrl  \
2694   https://cdn.sofifa.com/players/206/427/21_60.png   
4026   https://cdn.sofifa.com/players/243/019/21_60.png   
5235   https://cdn.sofifa.com/players/214/993/21_60.png   
18689  https://cdn.sofifa.com/players/256/442/21_60.png   
10221  https://cdn.sofifa.com/players/246/946/21_60.png   

                                               playerUrl Nationality  Age  \
2694   http://sofifa.com/player/206427/bogdan-butko/2...     Ukraine   29   
4026   http://sofifa.com/player/243019/tudor-baluta/2...     Romania   21   
5235   http://sofifa.com/player/214993/pablo-alvarado...   Argentina   34   
18689  http://sofifa.com/player/256442/jay-matete/210...     England   19   
10221  http://sofifa.com/player/246946/ismaila-coulib...        Mali   19   

       ↓OVA  POT                             Club  ...     A/W     D/W   IR  \
2694     73   73              \n\n\n\nLech Poznań  ...    High  Medium  1 ★   
4026     71   82   \n\n\n\nBrighton & Hove Albion  ...  Medium    High  1 ★   
5235     70   70  \n\n\n\nIndependiente del Valle  ...  Medium  Medium  1 ★   
18689    51   63           \n\n\n\nFleetwood Town  ...  Medium  Medium  1 ★   
10221    65   77             \n\n\n\nBeerschot AC  ...    High     Low  1 ★   

      PAC SHO  PAS DRI DEF PHY Hits  
2694   77  48   65  68  69  68    2  
4026   60  49   57  49  69  74   71  
5235   51  36   50  61  70  72    3  
18689  65  43   51  56  39  46  NaN  
10221  77  62   59  67  58  63   23  

[5 rows x 77 columns]

### Note 1: We choose to work with dataset 2 because the teams and contract data are already separated and dataset 2 looks cleaner

In [289]:
df.isnull().sum()[:50]

ID                      0
Name                    0
LongName                0
photoUrl                0
playerUrl               0
Nationality             0
Age                     0
↓OVA                    0
POT                     0
Club                    0
Contract                0
Positions               0
Height                  0
Weight                  0
Preferred Foot          0
BOV                     0
Best Position           0
Joined                  0
Loan Date End       17966
Value                   0
Wage                    0
Release Clause          0
Attacking               0
Crossing                0
Finishing               0
Heading Accuracy        0
Short Passing           0
Volleys                 0
Skill                   0
Dribbling               0
Curve                   0
FK Accuracy             0
Long Passing            0
Ball Control            0
Movement                0
Acceleration            0
Sprint Speed            0
Agility                 0
Reactions   

### Preliminary Problem 1 : 
1. "Hits" column have 2595 null value
2. "Loan Date End" column have 17966 null value
3. '↓OVA" ad is not correctly named (contains a special character)
4. 'photoUrl' and 'playerUrl' are not necessary for our analysis, we will delete them

In [259]:
df.duplicated().sum()

0

In [291]:
df.shape

(18979, 77)

### Note 2 :  Preliminary information
1. Our dataset have 18979 lines and 77 columns
2. not duplicate data in our dataset
3. "Hits" column have 2595 null value
4. "Loan Date End" column have 17966 null value
5. ↓OVA ad is not correctly named (contains a special character)
6. 'photoUrl' and 'playerUrl' are not necessary for our analysis, we will delete them

## Cleaning our dataset

#### We save our dataset and clean a copy

In [292]:
# First, we copy the df to save an original version
cp_df1 = df1.copy()
cp_df2 = df2.copy()

#We start with the second dataset
df = df2.copy()

### Fix Preliminary Problem 1 : 

#### Drop the columns photoUrl, playerUrl

In [295]:
# Drop the columns photoUrl, playerUrl
df.drop(['photoUrl','playerUrl'],axis=1,inplace=True,errors='ignore')

#### rename '↓OVA' column to'OVA'

In [296]:
# rename '↓OVA' column to'OVA'
df.rename(columns={'↓OVA':'OVA'},inplace=True,)

#### visual evaluation of the first 13 columns

In [264]:
df.iloc[:,:13].head(5)

ID               Name                      LongName Nationality  Age  \
0  158023           L. Messi                  Lionel Messi   Argentina   33   
1   20801  Cristiano Ronaldo  C. Ronaldo dos Santos Aveiro    Portugal   35   
2  200389           J. Oblak                     Jan Oblak    Slovenia   27   
3  192985       K. De Bruyne               Kevin De Bruyne     Belgium   29   
4  190871          Neymar Jr    Neymar da Silva Santos Jr.      Brazil   28   

   OVA  POT                         Club     Contract   Positions Height  \
0   93   93         \n\n\n\nFC Barcelona  2004 ~ 2021  RW, ST, CF  170cm   
1   92   92             \n\n\n\nJuventus  2018 ~ 2022      ST, LW  187cm   
2   91   93      \n\n\n\nAtlético Madrid  2014 ~ 2023          GK  188cm   
3   91   91      \n\n\n\nManchester City  2015 ~ 2023     CAM, CM  181cm   
4   91   91  \n\n\n\nParis Saint-Germain  2017 ~ 2022     LW, CAM  175cm   

  Weight Preferred Foot  
0   72kg           Left  
1   83kg          Right  
2   87kg          Right  
3   70kg          Right  
4   68kg          Right

#### Problem 1:
- 'Club' column contains special characters "\n"

#### Solution to problem 1
- Remove all "\n" characters
- remove whitespace and characters from the beginning and the end

In [298]:
# Remove all special caractar in 'Club' Column using regex
df['Club'] = df['Club'].str.replace('\n', '', regex=True)

# Now, we remove whitespace and characters from the beginning and the end of 'Club'
df['Club'] = df['Club'].str.strip()

#### Evaluation of "Club" Column after clean it

In [299]:
df['Club'].unique()

array(['FC Barcelona', 'Juventus', 'Atlético Madrid', 'Manchester City',
       'Paris Saint-Germain', 'FC Bayern München', 'Liverpool',
       'Real Madrid', 'Chelsea', 'Tottenham Hotspur', 'Inter', 'Napoli',
       'Borussia Dortmund', 'Manchester United', 'Arsenal', 'Lazio',
       'Leicester City', 'Borussia Mönchengladbach', 'Real Sociedad',
       'Atalanta', 'Olympique Lyonnais', 'Milan', 'Villarreal CF',
       'RB Leipzig', 'Cagliari', 'Ajax', 'SL Benfica', 'AS Monaco',
       'Wolverhampton Wanderers', 'Everton', 'Fiorentina', 'FC Porto',
       'RC Celta', 'Torino', 'Sevilla FC', 'Grêmio', 'Real Betis', 'Roma',
       'Newcastle United', 'Eintracht Frankfurt', 'Valencia CF',
       'Medipol Başakşehir FK', 'Inter Miami', 'Bayer 04 Leverkusen',
       'Levante UD', 'Crystal Palace', 'Athletic Club de Bilbao',
       'Shanghai SIPG FC', 'VfL Wolfsburg',
       'Guangzhou Evergrande Taobao FC', 'Al Shabab',
       'Olympique de Marseille', 'Los Angeles FC',
       'Beijing Sino

In [300]:
df.iloc[:,:13].head(5)

ID               Name                      LongName Nationality  Age  \
0  158023           L. Messi                  Lionel Messi   Argentina   33   
1   20801  Cristiano Ronaldo  C. Ronaldo dos Santos Aveiro    Portugal   35   
2  200389           J. Oblak                     Jan Oblak    Slovenia   27   
3  192985       K. De Bruyne               Kevin De Bruyne     Belgium   29   
4  190871          Neymar Jr    Neymar da Silva Santos Jr.      Brazil   28   

   OVA  POT                 Club     Contract   Positions Height Weight  \
0   93   93         FC Barcelona  2004 ~ 2021  RW, ST, CF  170cm   72kg   
1   92   92             Juventus  2018 ~ 2022      ST, LW  187cm   83kg   
2   91   93      Atlético Madrid  2014 ~ 2023          GK  188cm   87kg   
3   91   91      Manchester City  2015 ~ 2023     CAM, CM  181cm   70kg   
4   91   91  Paris Saint-Germain  2017 ~ 2022     LW, CAM  175cm   68kg   

  Preferred Foot  
0           Left  
1          Right  
2          Right  
3          Right  
4          Right

####  problem 1 fixed

#### problem 2 : 
- Contract column have 2 informations : the start year and the end year

#### solution to problem 2 : 
- split it into 2 column : "Start Year and "End Year"
- remove all whitespace

In [301]:
# We split 'Contract' to 2 collumns, the first is the start year and the second is the End Year
df['Start Year'] = df['Contract'].str.split('~',expand=True)[0]
df['End Year'] = df['Contract'].str.split('~',expand=True)[1]

#### Evaluation of 'Start Year'

In [302]:
# Clean the 'Start Year'
df['Start Year'].value_counts()

2020                    5744
2019                    5219
2018                    2749
2017                    1604
2016                     881
Jun 30, 2021 On Loan     770
2015                     524
2014                     371
Free                     237
2013                     205
2012                     135
2011                     102
Dec 31, 2020 On Loan     100
2010                      61
2009                      42
2008                      32
May 31, 2021 On Loan      30
Dec 31, 2021 On Loan      22
2007                      19
Jan 31, 2021 On Loan      18
Jan 1, 2021 On Loan       15
Jun 30, 2022 On Loan      15
2005                      13
2006                      11
2004                       7
Nov 30, 2020 On Loan       7
Jun 23, 2021 On Loan       7
Nov 27, 2021 On Loan       6
2003                       6
Nov 22, 2020 On Loan       5
Jan 3, 2021 On Loan        3
2002                       3
Aug 31, 2021 On Loan       2
Jan 4, 2021 On Loan        2
May 31, 2022 O

#### problem 3
- Some start dates are more precise and contain the month, day and year while the majority only contain the year

#### Solution to problem 3 : 
- To keep the information accurate we will convert all start dates to the year-month-day format; we will do the same for the end date column. In this way, we will be able to calculate more accurately the duration of a player in a team (in year, month or day)

#### Problem 4 :
- We have **1013** players in Loan 
- We have **237** players Free

#### Solution to Problem 4 :
- We will create 2 more columns base on this result : 
    - "On Loan" column to indicate if a player is on loan or not
    - "Contract Status" column to indicate if a player is In contract or not

#### Create "On Loan" column to indicate if a player is on loan or not

In [305]:
# We create "On Loan" column to indicate if a player is on loan or not

# We set all value off On Loan column to 'No'
df['On Loan'] = 'No'

# We convert 'Start Year' column to string
pd.Series(['Start Year'], dtype="string")

# Now, we check if "On Loan" is in the value of 'Start Year' and set the value of 'On Loan' Column to Yes
for i in range(df.shape[0]):
    if 'On Loan' in df.iloc[i,75]:
        df.iloc[i,77] = 'Yes'

#### Evaluation of  "On Loan" column

In [306]:
df['On Loan'].value_counts()

No     17966
Yes     1013
Name: On Loan, dtype: int64

#### Create "Contract Status" column to indicate if a player is In contract or not

In [309]:
# We create "Contract Status" column to indicate if a player is In contract or not

# We set all value off 'Contract Status' column to 'In contract'
df['Contract Status'] = 'In contract'

# We convert 'Start Year' column to string
pd.Series(['Start Year'], dtype="string")

# Now, we check if "Free" is in the value of 'Start Year' and set the value of 'Contract Status' Column to Free
for i in range(df.shape[0]):
    if 'Free' in df.iloc[i,75]:
        df.iloc[i,78] = 'Free'

#### Evaluation of "Contract Status" column

In [310]:
df['Contract Status'].value_counts()

In contract    18742
Free             237
Name: Contract Status, dtype: int64

#### Problem 4 fixed, now, we fix problem 3

In [311]:
# Now we Replace 'On Loan' in the 'Start Year' by whitespace
df['Start Year'] = df['Start Year'].str.replace('On Loan','')

In [312]:
# Now, we remove whitespace and characters from the beginning and the end of 'Start Year'
df['Start Year'] = df['Start Year'].str.strip()

In [313]:
df['Start Year'][200:206]

200            2020
201            2020
202            2015
203            2018
204            2007
205    Jun 30, 2021
Name: Start Year, dtype: object

In [314]:
# Replace month name in column 'Start Year' by the number of the month 
df['Start Year'] = df['Start Year'].astype('str').map(lambda x : x.replace("Jan","01/"))
df['Start Year'] = df['Start Year'].astype('str').map(lambda x : x.replace("Feb","02/"))
df['Start Year'] = df['Start Year'].astype('str').map(lambda x : x.replace("Mar","03/"))
df['Start Year'] = df['Start Year'].astype('str').map(lambda x : x.replace("Apr","04/"))
df['Start Year'] = df['Start Year'].astype('str').map(lambda x : x.replace("May","05/"))
df['Start Year'] = df['Start Year'].astype('str').map(lambda x : x.replace("Jun","06/"))
df['Start Year'] = df['Start Year'].astype('str').map(lambda x : x.replace("Jul","07/"))
df['Start Year'] = df['Start Year'].astype('str').map(lambda x : x.replace("Aug","08/"))
df['Start Year'] = df['Start Year'].astype('str').map(lambda x : x.replace("Sep","09/"))
df['Start Year'] = df['Start Year'].astype('str').map(lambda x : x.replace("Oct","10/"))
df['Start Year'] = df['Start Year'].astype('str').map(lambda x : x.replace("Nov","11/"))
df['Start Year'] = df['Start Year'].astype('str').map(lambda x : x.replace("Dec","12/"))

# Replace ',' in column 'Start Year' by '/'
df['Start Year'] = df['Start Year'].astype('str').map(lambda x : x.replace(",","/"))

# Replace whitespace in column 'Start Year' by ''
df['Start Year'] = df['Start Year'].str.replace(' ', '', regex=True)

In [315]:
df['Start Year'][200:206]

200          2020
201          2020
202          2015
203          2018
204          2007
205    06/30/2021
Name: Start Year, dtype: object

In [316]:
df.iloc[200:210,:]

ID            Name                 LongName         Nationality  Age  \
200  199914           Allan   Allan Marques Loureiro              Brazil   29   
201  181872        A. Vidal             Arturo Vidal               Chile   33   
202  180930        E. Džeko               Edin Džeko  Bosnia Herzegovina   34   
203  178518   R. Nainggolan         Radja Nainggolan             Belgium   32   
204  176676         Marcelo  Marcelo Vieira da Silva              Brazil   32   
205  173731         G. Bale              Gareth Bale               Wales   30   
206  178562       E. Banega              Éver Banega           Argentina   32   
207  172871   J. Vertonghen           Jan Vertonghen             Belgium   33   
208  163705     S. Mandanda           Steve Mandanda              France   35   
209   41236  Z. Ibrahimović       Zlatan Ibrahimović              Sweden   38   

     OVA  POT                    Club              Contract     Positions  \
200   83   83                 Everton           2020 ~ 2023       CM, CDM   
201   83   83                   Inter           2020 ~ 2022  CM, CDM, CAM   
202   83   83                    Roma           2015 ~ 2022            ST   
203   83   83                   Inter           2018 ~ 2022   CAM, CM, CF   
204   83   83             Real Madrid           2007 ~ 2022            LB   
205   83   83       Tottenham Hotspur  Jun 30, 2021 On Loan            RW   
206   83   83               Al Shabab           2020 ~ 2023       CM, CDM   
207   83   83              SL Benfica           2020 ~ 2023        CB, LB   
208   83   83  Olympique de Marseille           2017 ~ 2024            GK   
209   83   83                   Milan           2020 ~ 2021            ST   

     ... SHO PAS DRI  DEF PHY Hits  Start Year End Year On Loan  \
200  ...  71  77  82   85  83  193        2020     2023      No   
201  ...  79  79  78   83  86  128        2020     2022      No   
202  ...  84  71  75   46  73   78        2015     2022      No   
203  ...  80  80  83   83  81  137        2018     2022      No   
204  ...  73  80  88   76  76  133        2007     2022      No   
205  ...  84  82  80   58  73  282  06/30/2021     None     Yes   
206  ...  74  84  82   76  70   61        2020     2023      No   
207  ...  62  72  71   85  78   71        2020     2023      No   
208  ...  81  78  84   48  81   43        2017     2024      No   
209  ...  88  77  79   34  77  425        2020     2021      No   

    Contract Status  
200     In contract  
201     In contract  
202     In contract  
203     In contract  
204     In contract  
205     In contract  
206     In contract  
207     In contract  
208     In contract  
209     In contract  

[10 rows x 79 columns]

**Problem 3.1:**
> Some "Start Year" values is in format "yyyy", some in format "mm/dd/yyyy" and some others in format "mm/d/yyyy" ; To resolve this, we need to create 3 different regular expression

In [317]:
df['Start Year'].value_counts()

2020          5744
2019          5219
2018          2749
2017          1604
2016           881
06/30/2021     770
2015           524
2014           371
Free           237
2013           205
2012           135
2011           102
12/31/2020     100
2010            61
2009            42
2008            32
05/31/2021      30
12/31/2021      22
2007            19
01/31/2021      18
06/30/2022      15
01/1/2021       15
2005            13
2006            11
06/23/2021       7
11/30/2020       7
2004             7
2003             6
11/27/2021       6
11/22/2020       5
01/3/2021        3
2002             3
05/31/2022       2
08/31/2021       2
01/4/2021        2
07/31/2021       1
12/30/2020       1
1998             1
12/30/2021       1
08/1/2021        1
06/30/2023       1
07/1/2021        1
07/5/2021        1
01/30/2021       1
01/17/2021       1
Name: Start Year, dtype: int64

In [319]:
# We use regular expression to work on "Start Year" to tranform it to "Start date"
year_exp = re.compile(r"([0-9]{4})")
date_exp1 = re.compile(r"([0-9]{2}/[0-9]{2}/[0-9]{4})")
date_exp2 = re.compile(r"([0-9]{2}/[0-9]{1}/[0-9]{4})")

In [320]:
print(date_exp1.fullmatch('06/12/2021'))

<re.Match object; span=(0, 10), match='06/12/2021'>


**Remark 1 :**
>for players who is free, we affected the value '03/13/2040' to "start_date" and "end Date" columns, just to have the same data type in this columns

In [321]:
# Now, we correct the value of "Start Year" Column
for i in range(df.shape[0]):
    if year_exp.match(df.iloc[i,75]):
        df.iloc[i,75] = '01/01/' + df.iloc[i,75]
    elif date_exp1.match(df.iloc[i,75]):
        df.iloc[i,75] = df.iloc[i,75]
    elif date_exp2.match(df.iloc[i,75]):
        df.iloc[i,75] = df.iloc[i,75].split('/')[0] + '/0' + df.iloc[i,75].split('/')[1]  + df.iloc[i,75].split('/')[2]
    elif df.iloc[i,75] == 'Free':
        # for players who is free, we affected the value '03/13/2040' to "start_date" column, just to have the same data type in this columns
        df.iloc[i,75] = '03/13/2040'

In [323]:
df['Start Year'].value_counts()

01/01/2020    5744
01/01/2019    5219
01/01/2018    2749
01/01/2017    1604
01/01/2016     881
06/30/2021     770
01/01/2015     524
01/01/2014     371
03/13/2040     237
01/01/2013     205
01/01/2012     135
01/01/2011     102
12/31/2020     100
01/01/2010      61
01/01/2009      42
01/01/2008      32
05/31/2021      30
12/31/2021      22
01/01/2007      19
01/31/2021      18
06/30/2022      15
01/012021       15
01/01/2005      13
01/01/2006      11
01/01/2004       7
11/30/2020       7
06/23/2021       7
11/27/2021       6
01/01/2003       6
11/22/2020       5
01/032021        3
01/01/2002       3
05/31/2022       2
01/042021        2
08/31/2021       2
01/17/2021       1
01/01/1998       1
01/30/2021       1
08/012021        1
07/012021        1
06/30/2023       1
12/30/2020       1
12/30/2021       1
07/052021        1
07/31/2021       1
Name: Start Year, dtype: int64

#### Problem 3.2:  Some "Start Year" values in in format 'mm/ddyyyy', We reformat them to format 'mm/dd/yyy'

In [325]:
# Some "Start Year" values in in format 'mm/ddyyyy', We reformat them to format 'mm/dd/yyy'
date_exp3 =  re.compile(r"([0-9]{2}/[0-9]{6})")

date_exp3.match('07/052021')

<re.Match object; span=(0, 9), match='07/052021'>

In [326]:
# Now, we correct the value of "Start Year" Column
for i in range(df.shape[0]):
    if date_exp3.match(df.iloc[i,75]):
        df.iloc[i,75] = df.iloc[i,75].split('/')[0] + '/' + df.iloc[i,75].split('/')[1][:2] + '/' + df.iloc[i,75].split('/')[1][-4:]

In [327]:
# Now, we can easily convert it to correct date format "YYYY-MM-DD"
df['Start Year'] = pd.to_datetime(df['Start Year'], format='%m/%d/%Y')

In [329]:
df['Start Year'].value_counts()

2020-01-01    5744
2019-01-01    5219
2018-01-01    2749
2017-01-01    1604
2016-01-01     881
2021-06-30     770
2015-01-01     524
2014-01-01     371
2040-03-13     237
2013-01-01     205
2012-01-01     135
2011-01-01     102
2020-12-31     100
2010-01-01      61
2009-01-01      42
2008-01-01      32
2021-05-31      30
2021-12-31      22
2007-01-01      19
2021-01-31      18
2021-01-01      15
2022-06-30      15
2005-01-01      13
2006-01-01      11
2020-11-30       7
2004-01-01       7
2021-06-23       7
2021-11-27       6
2003-01-01       6
2020-11-22       5
2021-01-03       3
2002-01-01       3
2022-05-31       2
2021-08-31       2
2021-01-04       2
2023-06-30       1
2021-01-17       1
2021-01-30       1
2021-08-01       1
2021-07-05       1
2021-07-31       1
2020-12-30       1
2021-07-01       1
1998-01-01       1
2021-12-30       1
Name: Start Year, dtype: int64

#### Problem 3 fixed.

#### Evaluation of 'End Year' column

In [330]:
df['End Year'].isna().sum()

1250

In [331]:
df['End Year'].value_counts()

 2021    5959
 2022    4244
 2023    3418
 2024    2206
 2020    1289
 2025     606
 2026       5
 2028       1
 2027       1
Name: End Year, dtype: int64

#### Problem 5:
- "End Year" column have 1250 null values

In [332]:
df[df['End Year'].isna()].head()

ID            Name                          LongName Nationality  \
205  173731         G. Bale                       Gareth Bale       Wales   
248  193105       A. Areola                   Alphonse Areola      France   
254  200888  Danilo Pereira         Danilo Luís Hélio Pereira    Portugal   
289  230347  Welington Dano    Welington Kauê Dano Nascimento      Brazil   
292  230225  Juiano Mestres  Juan Everton Mestres de Mesquita      Brazil   

     Age  OVA  POT                 Club              Contract Positions  ...  \
205   30   83   83    Tottenham Hotspur  Jun 30, 2021 On Loan        RW  ...   
248   27   82   86               Fulham  Jun 30, 2021 On Loan        GK  ...   
254   28   82   82  Paris Saint-Germain  Jun 30, 2021 On Loan   CDM, CM  ...   
289   20   81   81              No Club                  Free    LB, LM  ...   
292   24   81   81              No Club                  Free   CB, CDM  ...   

    SHO PAS DRI  DEF PHY Hits Start Year End Year On Loan Contract Status  
205  84  82  80   58  73  282 2021-06-30     None     Yes     In contract  
248  79  80  84   55  78   94 2021-06-30     None     Yes     In contract  
254  63  72  73   80  85  285 2021-06-30     None     Yes     In contract  
289  54  76  78   75  74  172 2040-03-13     None      No            Free  
292  61  57  62   82  80   75 2040-03-13     None      No            Free  

[5 rows x 79 columns]

#### Remark 2:
- We have no information on the start and end date of the contract of players on loan. Indeed, when a player is under contract with club A, this club can loan him to another club B. The real starting date of the player's contract is the one with his home club, i.e. club A, and not the starting date of the loan with club B. Similarly, the true end date of the player's contract is the date with club A and not the loan end date.

- As we do not have the 'Start date' and 'End date' information for players on loan, we will assume that these dates are equal to the loan start date and loan end date respectively.

- Especially since we notice that the number of players on loan (1113) + the number of free players (237) = number of players without a contract end date (1250).

- So we don't need the column "Loan Date End" anymore, which we will delete after assigning its values to the column "End Year".

In [333]:
# Now, we correct the value of "End Year" Column
for i in range(df.shape[0]):
    if year_exp.match(str(df.iloc[i,76]).replace(' ','')):
        # If "End Year" is in format yyyy, we convert it to format mm/dd/yyyy , we consider the 31/12 ( the last day of the year)
        df.iloc[i,76] = '01/01/' + str(df.iloc[i,76]).replace(' ','')
    if df.iloc[i,78] == 'Free' :
        #i fplayer is free, Start Year = End Year = '03/13/2040' ( this is a random date, just use to identify this player)
        df.iloc[i,76] = '03/13/2040'
    if df.iloc[i,77] =='Yes' :
        # If player is in Loan, we copy the "Loan Date End" to "End Year"
        df.iloc[i,76] = str(df.iloc[i,16])

In [334]:
# Replace month name in column 'End Year' by the number of the month 
df['End Year'] = df['End Year'].astype('str').map(lambda x : x.replace("Jan","01/"))
df['End Year'] = df['End Year'].astype('str').map(lambda x : x.replace("Feb","02/"))
df['End Year'] = df['End Year'].astype('str').map(lambda x : x.replace("Mar","03/"))
df['End Year'] = df['End Year'].astype('str').map(lambda x : x.replace("Apr","04/"))
df['End Year'] = df['End Year'].astype('str').map(lambda x : x.replace("May","05/"))
df['End Year'] = df['End Year'].astype('str').map(lambda x : x.replace("Jun","06/"))
df['End Year'] = df['End Year'].astype('str').map(lambda x : x.replace("Jul","07/"))
df['End Year'] = df['End Year'].astype('str').map(lambda x : x.replace("Aug","08/"))
df['End Year'] = df['End Year'].astype('str').map(lambda x : x.replace("Sep","09/"))
df['End Year'] = df['End Year'].astype('str').map(lambda x : x.replace("Oct","10/"))
df['End Year'] = df['End Year'].astype('str').map(lambda x : x.replace("Nov","11/"))
df['End Year'] = df['End Year'].astype('str').map(lambda x : x.replace("Dec","12/"))

# Replace ',' in column 'End Year' by '/'
df['End Year'] = df['End Year'].astype('str').map(lambda x : x.replace(",","/"))

# Replace whitespace in column 'End Year' by ''
df['End Year'] = df['End Year'].str.replace(' ', '', regex=True)

In [338]:
# Now, we correct the value of "End Year" Column
for i in range(df.shape[0]):
    if date_exp2.match(df.iloc[i,76]):
        df.iloc[i,76] = df.iloc[i,76].split('/')[0] + '/0' + df.iloc[i,76].split('/')[1]  + df.iloc[i,76].split('/')[2]

In [339]:
df['End Year'].value_counts()

01/01/2021    5959
01/01/2022    4244
01/01/2023    3418
01/01/2024    2206
01/01/2020    1289
06/30/2021     770
01/01/2025     606
03/13/2040     237
12/31/2020     100
05/31/2021      30
12/31/2021      22
01/31/2021      18
01/012021       15
06/30/2022      15
06/23/2021       7
11/30/2020       7
11/27/2021       6
11/22/2020       5
01/01/2026       5
01/032021        3
08/31/2021       2
01/042021        2
05/31/2022       2
01/01/2028       1
01/30/2021       1
12/30/2020       1
01/01/2027       1
07/052021        1
07/31/2021       1
07/012021        1
12/30/2021       1
06/30/2023       1
08/012021        1
01/17/2021       1
Name: End Year, dtype: int64

#### Problem 5.2:  Some "End Year" values in in format 'mm/ddyyyy', We reformat them to format 'mm/dd/yyy'

In [340]:
# Now, we correct the value of "Start Year" Column
for i in range(df.shape[0]):
    if date_exp3.match(df.iloc[i,76]):
        df.iloc[i,76] = df.iloc[i,76].split('/')[0] + '/' + df.iloc[i,76].split('/')[1][:2] + '/' + df.iloc[i,76].split('/')[1][-4:]

In [341]:
# Now, we can easily convert it to correct date format "YYYY-MM-DD"
df['End Year'] = pd.to_datetime(df['End Year'], format='%m/%d/%Y')

In [342]:
df['End Year'].value_counts()

2021-01-01    5974
2022-01-01    4244
2023-01-01    3418
2024-01-01    2206
2020-01-01    1289
2021-06-30     770
2025-01-01     606
2040-03-13     237
2020-12-31     100
2021-05-31      30
2021-12-31      22
2021-01-31      18
2022-06-30      15
2021-06-23       7
2020-11-30       7
2021-11-27       6
2020-11-22       5
2026-01-01       5
2021-01-03       3
2021-01-04       2
2021-08-31       2
2022-05-31       2
2021-07-01       1
2028-01-01       1
2027-01-01       1
2021-08-01       1
2021-12-30       1
2023-06-30       1
2021-07-31       1
2021-07-05       1
2021-01-30       1
2021-01-17       1
2020-12-30       1
Name: End Year, dtype: int64

#### Problem 5 fixed.


#### Problem 6: Rename "Start Year","End Year" and delete "Contract" and 'Loan Date End'
- It is ok for "Start Year", we can rename it "Start Date"
- It is ok for "End Year", we can rename it "End Date"
- We can drop "Contract" and 'Loan Date End' column

In [343]:
df.rename(columns={"Start Year" :"Start Date", "End Year" :"End Date"},inplace=True)

In [344]:
df.drop(columns=['Loan Date End','Contract'],inplace=True, errors='ignore')

In [345]:
df.iloc[:,:13].head(5)

ID               Name                      LongName Nationality  Age  \
0  158023           L. Messi                  Lionel Messi   Argentina   33   
1   20801  Cristiano Ronaldo  C. Ronaldo dos Santos Aveiro    Portugal   35   
2  200389           J. Oblak                     Jan Oblak    Slovenia   27   
3  192985       K. De Bruyne               Kevin De Bruyne     Belgium   29   
4  190871          Neymar Jr    Neymar da Silva Santos Jr.      Brazil   28   

   OVA  POT                 Club   Positions Height Weight Preferred Foot  BOV  
0   93   93         FC Barcelona  RW, ST, CF  170cm   72kg           Left   93  
1   92   92             Juventus      ST, LW  187cm   83kg          Right   92  
2   91   93      Atlético Madrid          GK  188cm   87kg          Right   91  
3   91   91      Manchester City     CAM, CM  181cm   70kg          Right   91  
4   91   91  Paris Saint-Germain     LW, CAM  175cm   68kg          Right   91

#### Problem 6 fixed.

#### Evaluation of "Positions"

In [346]:
df['Positions'].value_counts()

CB              2441
GK              2074
ST              1864
CDM, CM          842
CM               795
                ... 
RB, RW, LB         1
CDM, LB, CB        1
CM, LW, ST         1
RWB, RW, RB        1
CDM, CAM, LM       1
Name: Positions, Length: 640, dtype: int64

In [347]:
df['Positions'].unique()

array(['RW, ST, CF', 'ST, LW', 'GK', 'CAM, CM', 'LW, CAM', 'ST', 'RW',
       'ST, LW, RW', 'CB', 'LW', 'CDM', 'CF, ST', 'LW, RW', 'CDM, CM',
       'CDM, RB', 'CF, CAM', 'LW, ST', 'CM', 'ST, CF, LW', 'RM, LM, CAM',
       'RB', 'RW, CAM, CM', 'LB', 'LM, CF', 'CF', 'RW, LW', 'CAM, RM, RW',
       'CM, CDM', 'CAM, CF, ST', 'CM, CDM, CAM', 'CF, LW, CAM',
       'CAM, RM, CF', 'LM, ST', 'RM, LM, RW', 'LM', 'CAM, RW', 'CB, CDM',
       'RW, RM', 'LW, CF', 'CM, RM, LM', 'LB, LM', 'CAM, CM, RM',
       'CAM, CM, CF', 'CAM, CF', 'LM, RM, LW', 'LM, LB, CM', 'CM, LM, LB',
       'RM, RW', 'RM, CM', 'CAM, CM, LW', 'CB, LB', 'RM, RB', 'ST, RW',
       'LM, RW, LW', 'RB, LB', 'RB, RM', 'RM', 'LM, RM, CF', 'CAM, RM',
       'RB, RWB', 'CDM, CB, CM', 'CAM, RM, ST', 'LM, LW, RM', 'CM, CAM',
       'ST, RM, CF', 'LM, RM', 'RM, CF', 'LM, LWB', 'RW, RM, CF',
       'RB, CM', 'LW, CAM, RW', 'CAM, LW, CM', 'CM, CAM, CDM',
       'RW, LW, CAM', 'CM, CAM, LM', 'CM, RM, ST', 'CDM, CM, RB',
       'ST, CAM', 

#### Problem 7 :
- Some players have multiple positions, sometimes they have the same positions except they are not listed in the same order, so we cannot group players with multiple positions in common.

#### Solution to Problem 7 :
- Sort player positions alphabetically

In [350]:
positions_ord = []
for p in range(len(df['Positions'])):
    po = sorted(df['Positions'][p].split(" ")) #ordered the list of positions
    po_join = ' '.join(po)
    positions_ord.append(po_join)

In [351]:
df['Positions'] = positions_ord
df['Positions'] = df['Positions'].astype(object)
df['Positions'].value_counts()

CB             2441
GK             2074
ST             1864
CDM, CM         842
CM              795
               ... 
CDM CM, LB,       1
CAM, LW RM,       1
RB RW, RWB,       1
LM, LWB, RB       1
LWB RB,           1
Name: Positions, Length: 470, dtype: int64

#### Problem 6 fixed : Note that, Now, we have 470 differents positions instead of 640 before cleaning

In [362]:
df.iloc[:,:13].head(5)

ID               Name                      LongName Nationality  Age  \
0  158023           L. Messi                  Lionel Messi   Argentina   33   
1   20801  Cristiano Ronaldo  C. Ronaldo dos Santos Aveiro    Portugal   35   
2  200389           J. Oblak                     Jan Oblak    Slovenia   27   
3  192985       K. De Bruyne               Kevin De Bruyne     Belgium   29   
4  190871          Neymar Jr    Neymar da Silva Santos Jr.      Brazil   28   

   OVA  POT                 Club   Positions  Height in cm Weight  \
0   93   93         FC Barcelona  CF RW, ST,           170   72kg   
1   92   92             Juventus      LW ST,           187   83kg   
2   91   93      Atlético Madrid          GK           188   87kg   
3   91   91      Manchester City     CAM, CM           181   70kg   
4   91   91  Paris Saint-Germain     CAM LW,           175   68kg   

  Preferred Foot  BOV  
0           Left   93  
1          Right   92  
2          Right   91  
3          Right   91  
4          Right   91

#### Problem 8 :  "Height" Column
- We remove "cm" in value
- We transform "Height" column type to int
- We Rename "Height" to "Height in cm"

In [353]:
df['Height'] = df['Height'].str.replace('cm','')

In [354]:
df['Height'].value_counts()

180     1474
178     1250
185     1182
183     1148
175     1091
        ... 
6'5"       1
5'4"       1
5'6"       1
206        1
155        1
Name: Height, Length: 62, dtype: int64

#### Problem 8.2 : 
- Some values of "Height" column is in feet, and some in cm.

#### Solution to problem 8.2 :
- We decide to convert all value in cm
> So, Height in cm = feet * 30.48 + inches * 2.54

In [355]:
# We remove the " at the end of some values
df['Height'] = df['Height'].str.replace('"','')
df['Height'].value_counts()[32:]

197     58
198     41
164     31
163     19
199     17
201     13
162     13
6'0      8
160      7
200      7
5'11     6
161      6
6'2      4
5'10     4
202      4
6'3      4
5'9      3
5'7      3
6'1      3
158      3
203      2
159      2
6'4      2
157      1
5'6      1
5'4      1
156      1
155      1
6'5      1
206      1
Name: Height, dtype: int64

In [357]:
feet_re = re.compile(r"([0-9]{1}'[0-9]{1})")
# feet_re.match("5'5")
# ch ="5'5"
# print(float(ch.split("'")[0]) * 30.48 + float('0.'+ ch.split("'")[1]) * 2.54)
# round(float(ch.split("'")[0]) * 30.48 + float('0.' + ch.split("'")[1]) * 2.54)

# # Some stackoverflow code . We use our own code here!
https://stackoverflow.com/questions/67314664/convert-feet-to-cm

A potentially faster way (vectorized) would be using str functions.

The operation you want to perform can be done by applying a dot-product between the 2 elements in your height column and the 2 conversion multipliers.
> [feet, inches] @ [30.48, 2.54] = feet*30.48 + inches*2.54


1. The str.split breaks the string into a list.
2. Then apply(pd.Series) breaks the list into 2 separate columns.
3. Finally, astype(int) converts each of the cells to int
4. Last, dot performs a dot product on each row with the conversion multipliers.

In [ ]:
# conversions = [30.48, 2.54]

# df['new_height'] =''

# for i in range(df.shape[0]):
#     if feet_re.match(df.iloc[i,9]) is None:
#         df.iloc[i,77] = df.iloc[i,9]
#     else:
#         df.iloc[i,77] = round(int(df.iloc[i,9].split("'")[0]) * 30.48 + int(df.iloc[i,9].split("'")[1]) * 2.54)

In [358]:
for i in range(df.shape[0]):
    if feet_re.match(df.iloc[i,9]):
        df.iloc[i,9] = round(int(df.iloc[i,9].split("'")[0]) * 30.48 + int(df.iloc[i,9].split("'")[1]) * 2.54)
    else:
        df.iloc[i,9] = df.iloc[i,9]

In [359]:
df['Height'].value_counts()

180    1474
178    1250
185    1182
183    1148
175    1091
       ... 
157       1
168       1
196       1
156       1
155       1
Name: Height, Length: 62, dtype: int64

In [360]:
df['Height'] = df['Height'].astype(int)

In [361]:
df.rename({'Height':'Height in cm'},inplace=True,axis=1)

In [363]:
df.iloc[:,:13].head(5)

ID               Name                      LongName Nationality  Age  \
0  158023           L. Messi                  Lionel Messi   Argentina   33   
1   20801  Cristiano Ronaldo  C. Ronaldo dos Santos Aveiro    Portugal   35   
2  200389           J. Oblak                     Jan Oblak    Slovenia   27   
3  192985       K. De Bruyne               Kevin De Bruyne     Belgium   29   
4  190871          Neymar Jr    Neymar da Silva Santos Jr.      Brazil   28   

   OVA  POT                 Club   Positions  Height in cm Weight  \
0   93   93         FC Barcelona  CF RW, ST,           170   72kg   
1   92   92             Juventus      LW ST,           187   83kg   
2   91   93      Atlético Madrid          GK           188   87kg   
3   91   91      Manchester City     CAM, CM           181   70kg   
4   91   91  Paris Saint-Germain     CAM LW,           175   68kg   

  Preferred Foot  BOV  
0           Left   93  
1          Right   92  
2          Right   91  
3          Right   91  
4          Right   91

#### Problem 8 fixed.

#### Problem 9 : "Weight" Column
- Values are not in the same unit, some contains caractars kg and is in kg and others contains lbs and is in lbs

#### Solution to Problem 9 :
- We remove "kg" 
- We delete spaces as a precaution
- We remove lbs in value and transform it to kg
- We Rename "Weight" to "Weight in kg"
> 1 pound (lb) is equal to 0.45359237 kilograms (kg)

In [366]:
df.Weight.value_counts()

70        1495
75        1457
80        1108
72        1022
78         991
          ... 
190lbs       1
185lbs       1
110          1
107          1
139lbs       1
Name: Weight, Length: 79, dtype: int64

In [367]:
df['Weight'] = df['Weight'].str.replace('kg','')
df['Weight'] = df['Weight'].str.replace(' ','')

In [370]:
weight_lb_re = re.compile(r"([0-9]{1,4}lbs)")

# w= '172lbs'
# if weight_lb_re.match(w):
#     print(round(float(w.replace('lbs','')) * 0.45359237) )

In [371]:
for i in range(df.shape[0]):
    if weight_lb_re.match(df.iloc[i,10]):
        df.iloc[i,10] = round(float(df.iloc[i,10].replace('lbs','')) * 0.45359237)
    else:
        df.iloc[i,10] = df.iloc[i,10]

In [372]:
df['Weight'].value_counts()

70     1495
75     1457
80     1108
72     1022
78      991
       ... 
79        1
74        1
110       1
107       1
53        1
Name: Weight, Length: 79, dtype: int64

In [375]:
df['Weight'] = df['Weight'].astype(int)
df.rename(columns = {'Weight':'Weight in Kg'},inplace=True)

In [376]:
df.iloc[:,:13].head(5)

ID               Name                      LongName Nationality  Age  \
0  158023           L. Messi                  Lionel Messi   Argentina   33   
1   20801  Cristiano Ronaldo  C. Ronaldo dos Santos Aveiro    Portugal   35   
2  200389           J. Oblak                     Jan Oblak    Slovenia   27   
3  192985       K. De Bruyne               Kevin De Bruyne     Belgium   29   
4  190871          Neymar Jr    Neymar da Silva Santos Jr.      Brazil   28   

   OVA  POT                 Club   Positions  Height in cm  Weight in Kg  \
0   93   93         FC Barcelona  CF RW, ST,           170            72   
1   92   92             Juventus      LW ST,           187            83   
2   91   93      Atlético Madrid          GK           188            87   
3   91   91      Manchester City     CAM, CM           181            70   
4   91   91  Paris Saint-Germain     CAM LW,           175            68   

  Preferred Foot  BOV  
0           Left   93  
1          Right   92  
2          Right   91  
3          Right   91  
4          Right   91

#### Problem 9 fixed.

#### Evaluation of 'Preferred Foot'

In [377]:
df['Preferred Foot'].value_counts()

Right    14445
Left      4534
Name: Preferred Foot, dtype: int64

####  'Preferred Foot' column is ok

#### Evaluation of 'BOV'

In [378]:
df['BOV'].value_counts()

66    1229
67    1213
65    1193
64    1136
68    1112
69    1019
63     983
70     975
62     859
71     836
72     789
61     768
73     685
60     651
74     565
59     540
75     530
76     398
58     389
57     357
77     354
56     290
78     275
55     263
54     222
53     185
79     185
80     173
52     151
81     134
82      94
51      89
83      73
50      55
84      50
85      43
86      29
87      25
49      22
88      11
48       9
89       8
91       5
90       5
92       1
93       1
Name: BOV, dtype: int64

####  'BOV'column is ok

#### Evaluation of columns 13 to 26

In [380]:
df.iloc[:,13:26].head(5)

Best Position        Joined    Value   Wage Release Clause  Attacking  \
0            RW   Jul 1, 2004  €103.5M  €560K        €138.4M        429   
1            ST  Jul 10, 2018     €63M  €220K         €75.9M        437   
2            GK  Jul 16, 2014    €120M  €125K        €159.4M         95   
3           CAM  Aug 30, 2015    €129M  €370K          €161M        407   
4            LW   Aug 3, 2017    €132M  €270K        €166.5M        408   

   Crossing  Finishing  Heading Accuracy  Short Passing  Volleys  Skill  \
0        85         95                70             91       88    470   
1        84         95                90             82       86    414   
2        13         11                15             43       13    109   
3        94         82                55             94       82    441   
4        85         87                62             87       87    448   

   Dribbling  
0         96  
1         88  
2         12  
3         88  
4         95

#### Evaluation of 'Best Position'

In [381]:
df['Best Position'].value_counts()

CB     3686
ST     2680
CAM    2299
GK     2075
RM     1611
CDM    1445
LB     1086
RB     1079
CM     1047
LM      871
RW      298
RWB     277
LWB     261
LW      186
CF       78
Name: Best Position, dtype: int64

#### is ok 'Best Position'

In [383]:
df.iloc[:7,13:26]

Best Position        Joined    Value   Wage Release Clause  Attacking  \
0            RW   Jul 1, 2004  €103.5M  €560K        €138.4M        429   
1            ST  Jul 10, 2018     €63M  €220K         €75.9M        437   
2            GK  Jul 16, 2014    €120M  €125K        €159.4M         95   
3           CAM  Aug 30, 2015    €129M  €370K          €161M        407   
4            LW   Aug 3, 2017    €132M  €270K        €166.5M        408   
5            ST   Jul 1, 2014    €111M  €240K          €132M        423   
6            RW   Jul 1, 2017  €120.5M  €250K        €144.3M        392   

   Crossing  Finishing  Heading Accuracy  Short Passing  Volleys  Skill  \
0        85         95                70             91       88    470   
1        84         95                90             82       86    414   
2        13         11                15             43       13    109   
3        94         82                55             94       82    441   
4        85         87                62             87       87    448   
5        71         94                85             84       89    407   
6        79         91                59             84       79    406   

   Dribbling  
0         96  
1         88  
2         12  
3         88  
4         95  
5         85  
6         90

#### Problem 10:
- "Joined" in not in the good date format

#### Solution to Problem 10:
- We convert it to date format "yyyy-mm-dd"

In [385]:
# Replace month name in column 'Joined' by the number of the month 
df['Joined'] = df['Joined'].astype('str').map(lambda x : x.replace("Jan","01/"))
df['Joined'] = df['Joined'].astype('str').map(lambda x : x.replace("Feb","02/"))
df['Joined'] = df['Joined'].astype('str').map(lambda x : x.replace("Mar","03/"))
df['Joined'] = df['Joined'].astype('str').map(lambda x : x.replace("Apr","04/"))
df['Joined'] = df['Joined'].astype('str').map(lambda x : x.replace("May","05/"))
df['Joined'] = df['Joined'].astype('str').map(lambda x : x.replace("Jun","06/"))
df['Joined'] = df['Joined'].astype('str').map(lambda x : x.replace("Jul","07/"))
df['Joined'] = df['Joined'].astype('str').map(lambda x : x.replace("Aug","08/"))
df['Joined'] = df['Joined'].astype('str').map(lambda x : x.replace("Sep","09/"))
df['Joined'] = df['Joined'].astype('str').map(lambda x : x.replace("Oct","10/"))
df['Joined'] = df['Joined'].astype('str').map(lambda x : x.replace("Nov","11/"))
df['Joined'] = df['Joined'].astype('str').map(lambda x : x.replace("Dec","12/"))

# Replace ',' in column 'Joined' by '/'
df['Joined'] = df['Joined'].astype('str').map(lambda x : x.replace(",","/"))

# Replace whitespace in column 'Joined' by ''
df['Joined'] = df['Joined'].str.replace(' ', '', regex=True)

In [386]:
# Now, we correct the value of "Joined" Column
for i in range(df.shape[0]):
    if date_exp2.match(df.iloc[i,14]):
        df.iloc[i,14] = df.iloc[i,14].split('/')[0] + '/0' + df.iloc[i,14].split('/')[1]  + df.iloc[i,14].split('/')[2]
    if date_exp3.match(df.iloc[i,14]):
        df.iloc[i,14] = df.iloc[i,14].split('/')[0] + '/' + df.iloc[i,14].split('/')[1][:2] + '/' + df.iloc[i,14].split('/')[1][-4:]

In [387]:
# Now, we can easily convert it to correct date format "YYYY-MM-DD"
df['Joined'] = pd.to_datetime(df['Joined'], format='%m/%d/%Y')

In [388]:
df['Joined'].value_counts()

2019-07-01    1344
2018-07-01     865
2019-01-01     682
2017-07-01     520
2020-07-01     435
              ... 
2012-09-04       1
2012-03-30       1
2014-02-04       1
2017-11-01       1
2017-08-27       1
Name: Joined, Length: 1869, dtype: int64

#### Problem 10 fixed :  It is good for "Joined" Column.

#### Evaluation of 'Value' columns

In [389]:
df.iloc[:7,13:26]

Best Position     Joined    Value   Wage Release Clause  Attacking  \
0            RW 2004-07-01  €103.5M  €560K        €138.4M        429   
1            ST 2018-07-10     €63M  €220K         €75.9M        437   
2            GK 2014-07-16    €120M  €125K        €159.4M         95   
3           CAM 2015-08-30    €129M  €370K          €161M        407   
4            LW 2017-08-03    €132M  €270K        €166.5M        408   
5            ST 2014-07-01    €111M  €240K          €132M        423   
6            RW 2017-07-01  €120.5M  €250K        €144.3M        392   

   Crossing  Finishing  Heading Accuracy  Short Passing  Volleys  Skill  \
0        85         95                70             91       88    470   
1        84         95                90             82       86    414   
2        13         11                15             43       13    109   
3        94         82                55             94       82    441   
4        85         87                62             87       87    448   
5        71         94                85             84       89    407   
6        79         91                59             84       79    406   

   Dribbling  
0         96  
1         88  
2         12  
3         88  
4         95  
5         85  
6         90

In [151]:
df['Value'].value_counts()[30:60]

€625K    236
€2.3M    222
€900K    218
€250K    214
€750K    210
€2M      204
€2.2M    184
€675K    180
€775K    180
€925K    168
€800K    167
€975K    165
€875K    163
€2.4M    160
€2.7M    157
€240K    154
€2.5M    153
€5.5M    149
€650K    139
€210K    137
€950K    133
€200K    127
€3M      125
€6.5M    122
€110K    121
€190K    120
€3.1M    117
€2.6M    116
€230K    116
€220K    116
Name: Value, dtype: int64

#### Problem 11 :  "Value" column
- Some values are in Millions and others in Thousands, 
- We will reduce them to the level of units

#### Solution Problem 11 :
- Remove the "€" symbole
- Remove the "M" and convert the value to units
- Remove the "K" and convert the value to units
- Convert column to int

> Convertion Formula
- value in units = value in Millions * 1.000.000
- value in units = value in Thousands * 1.000

In [390]:
df['Value'] = df['Value'].str.replace('€','')
df['Value'] = df['Value'].str.replace(' ','')

In [391]:
df['Value'].value_counts()

1.2M     582
1.1M     559
1.3M     534
1.6M     532
1M       497
        ... 
132M       1
45M        1
50M        1
71.5M      1
124M       1
Name: Value, Length: 255, dtype: int64

In [393]:
million_exp = re.compile(r"([0-9]{0,4}.*[0-9]{0,2}M)")

# million_exp.match("4M")
# ch = '63M'
# float(ch.replace('M','')) * 1000000

In [394]:
thousand_exp = re.compile(r"([0-9]{1,3}K)")

thousand_exp.match("925K")
ch = "325K"
int(ch.replace('K','')) * 1000

325000

In [395]:
# Now, we correct the value of "Value" Column
for i in range(df.shape[0]):
    if million_exp.match(df.iloc[i,15]):
        df.iloc[i,15] = float( df.iloc[i,15].replace('M','') ) * 1000000
    elif thousand_exp.match(df.iloc[i,15]):
        df.iloc[i,15] = float(df.iloc[i,15].replace('K','')) * 1000
    else:
        df.iloc[i,15] = float(df.iloc[i,15])

In [396]:
df['Value'].value_counts()

1200000.0     582
1100000.0     559
1300000.0     534
1600000.0     532
1000000.0     497
             ... 
63500000.0      1
55000000.0      1
45000000.0      1
95000000.0      1
61500000.0      1
Name: Value, Length: 255, dtype: int64

#### Problem 11 fixed : "Wage" and "Release Clause" Columns
- "Wage" and "Release Clause" Columns have the same problem like 'Value' column

#### Solution to Problem 12 and 13 : 
> We apply the same solution like 'Value' column

In [397]:
df['Wage'] = df2['Wage']

In [398]:
df['Wage'].value_counts()

€2K      2899
€500     2042
€1K      1891
€3K      1823
€4K      1050
         ... 
€270K       2
€290K       1
€370K       1
€88K        1
€560K       1
Name: Wage, Length: 134, dtype: int64

In [399]:
df['Wage'] = df['Wage'].str.replace('€','')
df['Wage'] = df['Wage'].str.replace(' ','')

In [400]:
# Now, we correct the value of "Wage" Column
for i in range(df.shape[0]):
    if million_exp.match(df.iloc[i,16]):
        df.iloc[i,16] = float( df.iloc[i,16].replace('M','') ) * 1000000
    elif thousand_exp.match(df.iloc[i,16]):
        df.iloc[i,16] = float(df.iloc[i,16].replace('K','')) * 1000
    else:
        df.iloc[i,16] = float(df.iloc[i,16])

In [401]:
df['Wage'].value_counts()

2000.0      2899
500.0       2042
1000.0      1891
3000.0      1823
4000.0      1050
            ... 
87000.0        2
370000.0       1
560000.0       1
88000.0        1
290000.0       1
Name: Wage, Length: 134, dtype: int64

In [402]:
df['Release Clause'].value_counts()

€0        1261
€1.1M      609
€1.2M      504
€1.3M      443
€1.4M      385
          ... 
€42M         1
€78.9M       1
€54.3M       1
€227K        1
€69.9M       1
Name: Release Clause, Length: 1216, dtype: int64

In [403]:
df['Release Clause'] = df['Release Clause'].str.replace('€','')
df['Release Clause'] = df['Release Clause'].str.replace(' ','')

In [404]:
# Now, we correct the value of "Release Clause" Column
for i in range(df.shape[0]):
    if million_exp.match(df.iloc[i,17]):
        df.iloc[i,17] = float( df.iloc[i,17].replace('M','') ) * 1000000
    elif thousand_exp.match(df.iloc[i,17]):
        df.iloc[i,17] = float(df.iloc[i,17].replace('K','')) * 1000
    else:
        df.iloc[i,17] = float(df.iloc[i,17])

In [405]:
df['Release Clause'].value_counts()

0.0           1261
1100000.0      609
1200000.0      504
1300000.0      443
1400000.0      385
              ... 
27500000.0       1
54200000.0       1
24500000.0       1
81900000.0       1
35000000.0       1
Name: Release Clause, Length: 1216, dtype: int64

#### Problem 12 and 12 fixed :  All is ok for "Value", "Wage" and "Release Clause"

In [407]:
df.iloc[:7,13:26]

Best Position     Joined        Value      Wage Release Clause  Attacking  \
0            RW 2004-07-01  103500000.0  560000.0    138400000.0        429   
1            ST 2018-07-10   63000000.0  220000.0     75900000.0        437   
2            GK 2014-07-16  120000000.0  125000.0    159400000.0         95   
3           CAM 2015-08-30  129000000.0  370000.0    161000000.0        407   
4            LW 2017-08-03  132000000.0  270000.0    166500000.0        408   
5            ST 2014-07-01  111000000.0  240000.0    132000000.0        423   
6            RW 2017-07-01  120500000.0  250000.0    144300000.0        392   

   Crossing  Finishing  Heading Accuracy  Short Passing  Volleys  Skill  \
0        85         95                70             91       88    470   
1        84         95                90             82       86    414   
2        13         11                15             43       13    109   
3        94         82                55             94       82    441   
4        85         87                62             87       87    448   
5        71         94                85             84       89    407   
6        79         91                59             84       79    406   

   Dribbling  
0         96  
1         88  
2         12  
3         88  
4         95  
5         85  
6         90

#### All is ok for columns 13 to 26.

#### Evaluation of columns 26 to 40

In [410]:
df.iloc[:7,26:40]

Curve  FK Accuracy  Long Passing  Ball Control  Movement  Acceleration  \
0     93           94            91            96       451            91   
1     81           76            77            92       431            87   
2     13           14            40            30       307            43   
3     85           83            93            92       398            77   
4     88           89            81            95       453            94   
5     79           85            70            88       407            77   
6     83           69            75            89       460            94   

   Sprint Speed  Agility  Reactions  Balance  Power  Shot Power  Jumping  \
0            80       91         94       95    389          86       68   
1            91       87         95       71    444          94       95   
2            60       67         88       49    268          59       78   
3            76       78         91       76    408          91       63   
4            89       96         91       83    357          80       62   
5            78       77         93       82    420          89       84   
6            92       91         92       91    393          80       69   

   Stamina  
0       72  
1       84  
2       41  
3       89  
4       81  
5       76  
6       85

#### All is ok for columns 26 to 40.

#### Evaluation of columns 40 to 54

In [412]:
df.iloc[:10,40:54]

Strength  Long Shots  Mentality  Aggression  Interceptions  Positioning  \
0        69          94        347          44             40           93   
1        78          93        353          63             29           95   
2        78          12        140          34             19           11   
3        74          91        408          76             66           88   
4        50          84        356          51             36           87   
5        86          85        391          81             49           94   
6        75          84        376          63             55           91   
7        78          14        140          27             11           13   
8        76          79        341          62             38           91   
9        78          10        171          43             22           11   

   Vision  Penalties  Composure  Defending  Marking  Standing Tackle  \
0      95         75         96         91       32               35   
1      82         84         95         84       28               32   
2      65         11         68         57       27               12   
3      94         84         91        186       68               65   
4      90         92         93         94       35               30   
5      79         88         88         96       35               42   
6      84         83         90        122       38               43   
7      66         23         65         50       15               19   
8      80         70         84        100       34               34   
9      70         25         70         48       25               13   

   Sliding Tackle  Goalkeeping  
0              24           54  
1              24           58  
2              18          437  
3              53           56  
4              29           59  
5              19           51  
6              41           62  
7              16          439  
8              32           42  
9              10          439

#### All is ok for columns 40 to 54.

#### Evaluation of columns 54 to 70

In [413]:
df.iloc[:10,54:70]

GK Diving  GK Handling  GK Kicking  GK Positioning  GK Reflexes  \
0          6           11          15              14            8   
1          7           11          15              14           11   
2         87           92          78              90           90   
3         15           13           5              10           13   
4          9            9          15              15           11   
5         15            6          12               8           10   
6         14           14           9              11           14   
7         86           88          85              91           89   
8         13            5           7              11            6   
9         88           85          88              88           90   

   Total Stats  Base Stats  W/F  SM     A/W     D/W   IR  PAC  SHO  PAS  DRI  
0         2231         466  4 ★  4★  Medium     Low  5 ★   85   92   91   95  
1         2221         464  4 ★  5★    High     Low  5 ★   89   93   81   89  
2         1413         489  3 ★  1★  Medium  Medium  3 ★   87   92   78   90  
3         2304         485  5 ★  4★    High    High  4 ★   76   86   93   88  
4         2175         451  5 ★  5★    High  Medium  5 ★   91   85   86   94  
5         2195         457  4 ★  4★    High  Medium  4 ★   78   91   78   85  
6         2211         470  3 ★  4★    High  Medium  3 ★   93   86   81   90  
7         1389         490  3 ★  1★  Medium  Medium  3 ★   86   88   85   89  
8         2147         466  4 ★  5★    High     Low  3 ★   96   86   78   91  
9         1442         484  4 ★  1★  Medium  Medium  3 ★   88   85   88   90

#### Problem 14 :  columns "W/F", "SM" and "IR"
- Values of this columns contains the symbole "★" , we can not calculate avegare, max , min or others claculations with this

#### Solution to Problem 14 : 
- Remove the "★" symbole and whitespace
- Convert column to int

In [414]:
cols = ["W/F","SM","IR"]

for col in cols:
    df[col] = df[col].str.replace("★","").replace(" ",'')
    df[col] = df[col].astype(int)

In [415]:
df.iloc[:10,54:70]

GK Diving  GK Handling  GK Kicking  GK Positioning  GK Reflexes  \
0          6           11          15              14            8   
1          7           11          15              14           11   
2         87           92          78              90           90   
3         15           13           5              10           13   
4          9            9          15              15           11   
5         15            6          12               8           10   
6         14           14           9              11           14   
7         86           88          85              91           89   
8         13            5           7              11            6   
9         88           85          88              88           90   

   Total Stats  Base Stats  W/F  SM     A/W     D/W  IR  PAC  SHO  PAS  DRI  
0         2231         466    4   4  Medium     Low   5   85   92   91   95  
1         2221         464    4   5    High     Low   5   89   93   81   89  
2         1413         489    3   1  Medium  Medium   3   87   92   78   90  
3         2304         485    5   4    High    High   4   76   86   93   88  
4         2175         451    5   5    High  Medium   5   91   85   86   94  
5         2195         457    4   4    High  Medium   4   78   91   78   85  
6         2211         470    3   4    High  Medium   3   93   86   81   90  
7         1389         490    3   1  Medium  Medium   3   86   88   85   89  
8         2147         466    4   5    High     Low   3   96   86   78   91  
9         1442         484    4   1  Medium  Medium   3   88   85   88   90

In [195]:
df['A/W'].value_counts()

Medium    12701
High       5288
Low         990
Name: A/W, dtype: int64

In [416]:
df['D/W'].value_counts()

Medium    13956
High       3297
Low        1726
Name: D/W, dtype: int64

#### All is ok for columns 54 to 70.

#### Evaluation of columns 70 to last column

In [418]:
df.iloc[:10,70:]

DEF  PHY  Hits Start Date   End Date On Loan Contract Status
0   38   65   771 2004-01-01 2021-01-01      No     In contract
1   35   77   562 2018-01-01 2022-01-01      No     In contract
2   52   90   150 2014-01-01 2023-01-01      No     In contract
3   64   78   207 2015-01-01 2023-01-01      No     In contract
4   36   59   595 2017-01-01 2022-01-01      No     In contract
5   43   82   248 2014-01-01 2023-01-01      No     In contract
6   45   75   246 2017-01-01 2023-01-01      No     In contract
7   51   91   120 2018-01-01 2024-01-01      No     In contract
8   39   76  1.6K 2018-01-01 2022-01-01      No     In contract
9   45   88   130 2014-01-01 2022-01-01      No     In contract

In [220]:
# df['Hits'] = df2['Hits']

In [419]:
df['Hits'].isnull().sum()

2595

In [420]:
df1['Hits'].isnull().sum()

0

In [422]:
df1['ID'].value_counts()

251698    2
177564    1
228757    1
216467    1
259128    1
         ..
248337    1
228058    1
256728    1
236246    1
231423    1
Name: ID, Length: 18978, dtype: int64

In [423]:
df['ID'].value_counts()

198657    1
216467    1
206222    1
254787    1
3467      1
         ..
258155    1
238305    1
255685    1
213725    1
231423    1
Name: ID, Length: 18979, dtype: int64

#### problem 15 : "Hits" column
- Our dataset contains 2595 null values for the 'Hits' column; 

**Remark:** The first dataset contains none null values in the "Hits" column. Since both datasets have the 'ID' column in common.

#### Solution to problem 15 : 
- Delete all columns in the first dataset except the 'ID' and 'Hits' columns
- use the "ID" column to combine the data from both datasets to retrieve the values from the "Hits" column

In [433]:
df1_cols = df1.columns.tolist()
print(len(df1_cols))
df1_cols.remove('ID')
df1_cols.remove('Hits')
print(len(df1_cols))

tmp = df1.drop(columns=df1_cols, axis=1,errors='ignore')
tmp.sample(5)

77
75


ID   Hits
13636  258464    \n1
13789  212682    \n5
18406  252728      3
3171   226085    \n5
517    240754  \n286

In [482]:
tmp.drop_duplicates(inplace=True)

In [483]:
tmp.shape

(18978, 2)

In [484]:
tmp[tmp['Hits'] == '\n']

Empty DataFrame
Columns: [ID, Hits]
Index: []

In [485]:
tmp['Hits'].unique()

array(['\n372', '\n344', '\n86', '\n163', '\n273', '\n182', '\n646',
       '\n79', '\n164', '\n170', '\n93', '\n131', '\n89', '\n90', '\n169',
       '\n187', '\n103', '\n317', '\n239', '\n56', '\n130', '\n229',
       '\n124', '\n172', '\n114', '\n46', '\n663', '\n479', '\n203',
       '\n198', '\n122', '\n565', '\n152', '\n140', '\n276', '\n209',
       '\n25', '\n136', '\n109', '\n139', '\n42', '\n99', '\n137',
       '\n166', '\n80', '\n104', '\n150', '\n112', '\n285', '\n133',
       '\n96', '\n69', '\n143', '\n100', '\n38', '\n83', '\n67', '\n207',
       '\n416', '\n315', '\n206', '\n424', '\n191', '\n454', '\n271',
       '\n249', '\n121', '\n246', '\n171', '\n148', '\n329', '\n70',
       '\n97', '\n73', '\n210', '\n82', '\n146', '\n62', '\n165', '\n129',
       '\n162', '\n35', '\n65', '\n47', '\n77', '\n106', '\n84', '\n1.3K',
       '\n310', '\n181', '\n364', '\n200', '\n141', '\n134', '\n259',
       '\n309', '\n367', '\n224', '\n211', '\n60', '\n176', '\n108',
       '\n

In [486]:
tmp['Hits'].isnull().sum()

0

In [492]:
tmp.shape

(18978, 2)

In [494]:
df.drop(columns=['Hits'],inplace=True, errors='ignore')
df.shape

(18994, 76)

In [495]:
df = df.merge(tmp,on='ID',how='left')
df.shape

(18994, 77)

In [496]:
df['Hits'].isnull().sum()

1

In [448]:
df['Hits'].value_counts()

\n1      3216
\n2      2453
\n3      1774
\n4      1236
1        1105
         ... 
\n188       1
\n256       1
\n367       1
\n257       1
\n280       1
Name: Hits, Length: 408, dtype: int64

#### Problem 15.2 : in "Hits" column
- Values contains special '\n' 

#### Solution to Problem 15.2 :
- Remove the "\n" symbole and convert the value to int

In [497]:
df['Hits'] = df['Hits'].str.replace('\n','')
df['Hits'].unique()

array(['372', '344', '86', '163', '273', '182', '164', '79', '646', '93',
       '170', '131', '89', '90', '169', '187', '103', '130', '317', '239',
       '56', '229', '46', '124', '172', '114', '209', '663', '479', '203',
       '198', '122', '565', '152', '276', '140', '25', '80', '109', '139',
       '42', '99', '137', '166', '136', '96', '133', '150', '112', '285',
       '104', '67', '100', '38', '83', '207', '69', '143', '271', '329',
       '416', '315', '206', '424', '191', '249', '148', '121', '171',
       '246', '62', '70', '454', '97', '65', '84', '73', '210', '82',
       '165', '129', '146', '35', '106', '47', '77', '162', '200', '134',
       '141', '364', '310', '1.3K', '309', '181', '224', '259', '60',
       '211', '367', '176', '52', '111', '108', '63', '59', '27', '91',
       '117', '54', '350', '312', '245', '286', '132', '291', '258',
       '233', '767', '288', '57', '365', '75', '157', '115', '142', '149',
       '227', '92', '53', '113', '71', '144', '123', '

In [498]:
df['Hits'].isnull().sum()

2596

#### Problem 15.2.2 : in "Hits" column
- Some values are in Thousands and others in units 
- We will convert all them to units

#### Solution to Problem 15.2.2 :
- Remove the "K" symbole and convert the value to units
- Convert column to int

> Convertion Formula : value in units = value in Thousands * 1.000

In [499]:
hits_thousands = re.compile(r"([0-9]{0,2}.*[0-9]{0,2})")

ch = '5K'
if hits_thousands.match(ch):
    print(float(ch.replace('K','')) * 1000)

5000.0


In [500]:
# Don't forgett that now, "Hits" column have the index 76
for i in range(df.shape[0]):
    if hits_thousands.match(str(df.iloc[i,76])):
        df.iloc[i,76] = float(str(df.iloc[i,76]).replace('K','')) * 1000
    else:
        df.iloc[i,76] = float(df.iloc[i,76])

In [505]:
df.drop_duplicates(inplace=True)
df.shape

(18979, 77)

In [512]:
df['Hits'].isna().sum()

0

#### We replace null values in 'Hits' Column with the mean

In [510]:
df.fillna(df.mean() ,inplace=True)

<ipython-input-510-5be9ac87e9f8>:1: FutureWarning: DataFrame.mean and DataFrame.median with numeric_only=None will include datetime64 and datetime64tz columns in a future version.
  df.fillna(df.mean() ,inplace=True)


In [511]:
df['Hits'].isna().sum()

0

In [502]:
df.iloc[:5,60:]

Base Stats  W/F  SM     A/W     D/W  IR  PAC  SHO  PAS  DRI  DEF  PHY  \
0         466    4   4  Medium     Low   5   85   92   91   95   38   65   
1         464    4   5    High     Low   5   89   93   81   89   35   77   
2         489    3   1  Medium  Medium   3   87   92   78   90   52   90   
3         485    5   4    High    High   4   76   86   93   88   64   78   
4         451    5   5    High  Medium   5   91   85   86   94   36   59   

  Start Date   End Date On Loan Contract Status      Hits  
0 2004-01-01 2021-01-01      No     In contract  372000.0  
1 2018-01-01 2022-01-01      No     In contract  344000.0  
2 2014-01-01 2023-01-01      No     In contract   86000.0  
3 2015-01-01 2023-01-01      No     In contract  163000.0  
4 2017-01-01 2022-01-01      No     In contract  273000.0

In [507]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 18979 entries, 0 to 18993
Data columns (total 77 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   ID                18979 non-null  int64         
 1   Name              18979 non-null  object        
 2   LongName          18979 non-null  object        
 3   Nationality       18979 non-null  category      
 4   Age               18979 non-null  int64         
 5   OVA               18979 non-null  int64         
 6   POT               18979 non-null  int64         
 7   Club              18979 non-null  category      
 8   Positions         18979 non-null  category      
 9   Height in cm      18979 non-null  int32         
 10  Weight in Kg      18979 non-null  int32         
 11  Preferred Foot    18979 non-null  object        
 12  BOV               18979 non-null  int64         
 13  Best Position     18979 non-null  category      
 14  Joined            1897

### Small Problem :
- We convert some columns to "category" 
>   categorical_cols = ['Nationality','Positions', 'Club','Best Position','A/W','D/W','On Loan','Contract Status']

In [514]:
categorical_cols = ['Nationality','Positions', 'Club','Best Position','A/W','D/W','On Loan','Contract Status']

for col in categorical_cols:
    df[col] = df[col].astype('category')

In [515]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 18979 entries, 0 to 18993
Data columns (total 77 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   ID                18979 non-null  int64         
 1   Name              18979 non-null  object        
 2   LongName          18979 non-null  object        
 3   Nationality       18979 non-null  category      
 4   Age               18979 non-null  int64         
 5   OVA               18979 non-null  int64         
 6   POT               18979 non-null  int64         
 7   Club              18979 non-null  category      
 8   Positions         18979 non-null  category      
 9   Height in cm      18979 non-null  int32         
 10  Weight in Kg      18979 non-null  int32         
 11  Preferred Foot    18979 non-null  object        
 12  BOV               18979 non-null  int64         
 13  Best Position     18979 non-null  category      
 14  Joined            1897

In [516]:
df['Value'] = df['Value'].astype(int)
df['Wage'] = df['Wage'].astype(int)
df['Release Clause'] = df['Release Clause'].astype(int)

## Our Dataset is clean
> We save it to CSV file

In [517]:
print('Finals dimensions of our dataset :' ,df.shape)
df.to_csv("fifa21 raw data v2_clean.csv", index=False, encoding='utf-8')
print('Next time, we will see DATA ANALYSIS')

Finals dimensions of our dataset : (18979, 77)
Next time, we will see DATA ANALYSIS
